src/eduid/webapp/idp/views2/other_device.py

```diff
@@ -0,0 +1,348 @@
+                    # used during this authn, albeit on the other device, device #2) should also get added to the
+                    # pending request here on device #1.
+                    _sso_credentials_used: List[AuthnData] = []
+                    _request_count = 0
+                    for this in state.device2.credentials_used:
+                        authn = AuthnData(cred_id=this.credential_id, timestamp=this.ts)
+                        _sso_credentials_used += [authn]
+                        if this.source == UsedWhere.REQUEST:
+                            ticket.pending_request.credentials_used[this.credential_id] = this.ts
+                            _request_count += 1
+
+                    # Create/update SSO session
+                    sso_session = record_authentication(
+                        ticket, state.eppn, sso_session, _sso_credentials_used, current_app.conf.sso_session_lifetime
+                    )
+
+                    current_app.logger.info(
+                        f'Transferred {_request_count} request credentials used to login ref {ticket.request_ref}, '
+                        f'and {len(_sso_credentials_used)} to SSO session {sso_session.session_id}'
+                    )
+
+                    current_app.logger.debug(f'Saving SSO session {sso_session}')
+                    current_app.sso_sessions.save(sso_session)
+
+                    current_app.stats.count('login_using_other_device_finished')
+                else:
+                    current_app.logger.info(f'Use other device: Incorrect response_code')
+                    current_app.stats.count('login_using_other_device_incorrect_code')
+                    state.bad_attempts += 1
+            else:
+                current_app.logger.info(f'Not validating response code for use other device in state {state.state}')
+                state.bad_attempts += 1
+
+            if state.state != OtherDeviceState.DENIED:
+                if state.bad_attempts >= current_app.conf.other_device_max_code_attempts:
+                    current_app.logger.info(f'Use other device: too many response code attempts')
+                    current_app.stats.count('login_using_other_device_denied')
+                    state.state = OtherDeviceState.DENIED
+
+            if not current_app.other_device_db.save(state):
+                current_app.logger.warning(f'Login using other device: Failed saving state {state}')
+                return error_response(message=IdPMsg.general_failure)
+        elif action is not None:
+            current_app.logger.error(f'Login using other device: Unknown action: {action}')
+            return error_response(message=IdPMsg.general_failure)
+    else:
+        age = int((now - state.expires_at).total_seconds())
+        current_app.logger.info(f'Use other device state is expired ({age} seconds)')
+
+    if state.state in [OtherDeviceState.NEW, OtherDeviceState.IN_PROGRESS, OtherDeviceState.LOGGED_IN]:
```

Comment:
Why do we show the QR code when state is anything but NEW?

Reply:
To avoid page content moving in the frontend. Agree that a better fix is to have the frontend render something else than the QR code there, of the same size.